Add a slide on generic lambdas, including explicit template parameters (#189)
hsf-training/cpluspluscourse: talk/morelanguage/lambda.tex

  \end{block}
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[11]{Capture list}
  \begin{block}{Capturing this}
    Inside a (non-static) member function, we can capture \mintinline{cpp}{this}.
  \end{block}
  \begin{block}{}
    \begin{cppcode*}{gobble=2}
      [this](...) { use(*this); };
      [&](...) { use(*this); };
      [&, this](...) { use(*this); };
      [=](...) { use(*this); };       // deprecated in C++20
      [=, this](...) { use(*this); }; // allowed in C++20
    \end{cppcode*}
    The object at \mintinline{cpp}{*this} is captured by reference.
  \end{block}
  \pause
  \begin{block}{}
    \begin{cppcode*}{gobble=2}
      [*this](...) { use(*this); };
      [&, *this](...) { use(*this); };
      [=, *this](...) { use(*this); };
    \end{cppcode*}
    Allowed since \cpp17: the object at \mintinline{cpp}{*this} is captured by value.
  \end{block}
  Details in  \href{https://www.nextptr.com/tutorial/ta1430524603/capture-this-in-lambda-expression-timeline-of-change}{this blog post}.
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[14]{Generic lambdas}

>   \begin{block}{Generic lambdas (aka. polymorphic lambdas)}

Reviewer: ```suggestion
  \begin{block}{Generic lambdas (aka.\ polymorphic lambdas)}
```
Just a nitpick. Latex needs to know that this doesn't end a sentence for best spacing.

    \begin{itemize}

>       \item The type of a lambda parameter may be \mintinline{cpp}{auto}.

Reviewer: ```suggestion
      \item The type of lambda parameters may be \mintinline{cpp}{auto}.
```

      \begin{cppcode*}{linenos=false,gobble=4}
        auto add = [](auto a, auto b) { return a + b; };
      \end{cppcode*}
      \item The generated \mintinline{cpp}{operator()} becomes a template function:
      \begin{cppcode*}{linenos=false,gobble=4}
        template <typename T, typename U>
        auto operator()(T a, U b) const { return a + b; }
      \end{cppcode*}
      \item The types of \mintinline{cpp}{a} and \mintinline{cpp}{b} may be different.
    \end{itemize}
  \end{block}

>   \begin{block}{Explicit template parameters}

Reviewer: ```suggestion
  \begin{block}{Explicit template parameters (\cpp20)}
```
I'm not sure this is legal in our latex template, but what I mean is to put the restriction in the box title.

Author: Yes, we can do that!
